### DISCUSSION TITLE
Offline Auth

### DISCUSSION THREAD

**Discussion by henwoods** [supabase-community/supabase-kt#559] [General]
Is there a best practice for handling token refresh when the user is offline. I want to allow access to the app in this case using PowerSync or similar, it seems like when the device is offline and the token is expired the user will not be retrieved. Should this be possible?

**Comment by jan-tennert** [comment]
So you obviously can't refresh your session when the user is offline. Do you want to something like caching data locally and let the user still see cached content?

**Comment by henwoods** [comment]
Yes, I have a local copy of the user data which will process any changes when the internet is available again. It works great until the session tries to refresh after a cold start which results in the client failing with the `SessionStatus.NetworkError`. At the moment I’ve hacked around that by persisting the user id, and if I get the network error but  that user id exists and there is a session in the auth session manager then I just continue. 

Doing that feels a little gross so I was wondering if you had any ideas 😄

**Reply by jan-tennert** [reply]
I can tell what I did in one of my [apps](https://github.com/jan-tennert/EinkaufszettelNext). Everything I fetch from Supabase gets stored directly in a local SQL Database (via SQLDelight), and the user only sees what's already cached so there is no separation from online or offline mode. Then if the `sessionStatus` is either `Authenticated` or `NetworkError` the app navigates to the main app screen and shows that you might be offline, but still continues to show local data. Any fetch requests will obviously fail, but that doesn't matter

**Reply by henwoods** [reply]
Ah, much the same as what I’m doing now.

I just noticed that it’s possible to get the user from the session manager in this state as well. Any reason `currentUserOrNull` couldn’t return that cached user?

Also it looks like the client is persistent at wanting to refresh the token despite the lack of connectivity, checking every 10 seconds. I’ll check your app code to see if there is a way to handle that.

EDIT: On that second point, I’ve realised it’s also possible something in my app is trying to poke Supabase regularly so if you aren’t experiencing the same persistent token refresh I’ll investigate further at my end.

Thanks!

**Reply by jan-tennert** [reply]
You can also increase the token refresh retry timeout in the config.
